### DISCUSSION TITLE
[IR] MLIR Transform Dialect Migration

### DISCUSSION THREAD

**Discussion by chhzh123** [cornell-zhang/allo#207] [Projects]
As the [MLIR Transform dialect](https://mlir.llvm.org/docs/Dialects/Transform/) becomes more mature, we plan to migrate our original facilities of decoupled customizations in the Allo/HCL dialect to this Transform dialect. Previously we had some attempts on this migration (see https://github.com/cornell-zhang/hcl-dialect/pull/124), but this was later removed as it increased the maintenance cost. We may need a more systematic way to include this dialect and rewrite the passes inside [LoopTransformations.cpp](https://github.com/cornell-zhang/allo/blob/main/mlir/lib/Transforms/LoopTransformations.cpp).

**Comment by chhzh123** [comment]
At the same time, we may want to migrate our implementation from memref/affine to tensor/linalg dialects

**Comment by kkkaishao** [comment]
Currently finished:
- Updated cpp standard to 23 (mainly for ranges library, template lambdas etc.)
- Refactored `allo.split` to `transform.allo.split`. It takes the original loop and produces the outer and inner loop handle after splitting.
- Refactored `allo.tile` to `transform.allo.tile`. It now takes arbitrary number of original loop handles (in the same perfect loop band) and corresponding tiling factors, producing tiled loops in [tile loop, point loop] bundles.
- Refactored `allo.reorder` to `transform.allo.reorder`. 
  - Breaking change: the input permutation map is the relative order of input loop handles. e.g. given loop nest `[L0, L1, L2]`, input loop handles `[L0, L2]` and permutation map `[1, 0]`, the operation will transform the loop nest into `[L2, L1, L0]`. Positions of unmentioned loop handles will be unchanged (which is a partial sort).
  - Added a permutability check using `affine::isValidLoopInterchangePermutation`, because sometimes the permutation attempt may violate data dependencies

**Reply by chhzh123** [reply]
Thanks for the update! Can you create a list of actions or a roadmap you want to take so we can iterate?

**Comment by kkkaishao** [comment]
Refactor roadmap:
- Compilation configurations (cpp standard, cmake configurations etc.)
- Refactor original transform ops (which take some op handles as input) to their corresponding transform dialect version.
  - Wrap the original transform logic with transform dialect interfaces.
  - Includes some enhancements like boundary check, validity check.
  - Fix some todos in original version to support more complicated use cases.
  - If the original op has a superior replacement in the upstream transform dialect, then use the upstream version. (only `allo.outline` satisfies this condition now).
  - The form of arguments and results of transform ops will be easy to construct and use as much as possible.
- Create a Python class `TSWriter` (a.k.a Transform Script Writer) to explicitly manage the transform script context.
  - The object will be add to the `Schedule` context.
  - Rewrite schedule methods implementations in `Schedule` (e.g. `.to()`, `.pipeline()`). They now only insert bundles of `transform.match` ops and corresponding transform ops to `TSWriter` (which will be just adding some strings and very easy to implement). `TSWriter` internally  composes a complete transform script that satisfies the requirements of transformation engine.
  - The top-level user interface will not be changed.
  - `TSWriter` will provide a `.commit()`-like API, which commits the entire transform sequence and apply them all in one-time.
  - Provide a failure handling mechanism.
 
What may be done further:
- Rewrite some analysis and transforms (which are currently written in Python) in C++ using transform dialect, since Python seems to have higher maintenance cost in large projects (and it has been a bit messy in some places now).
- Provide polyhedral analysis using ISL library, which may be useful in systolic array generation (AutoSA-like)
- Add a encoding attribute for array partition which can used in RankedTensorType values. e.g.
```
#partition = allo.memory_layout<{
    impl = bram,
    part_factor = {2, 4}
}>
%mat: tensor<16x16xi8, #partition>
```
Then we can unify ram interfaces between different modules using layout propagation with an algorithm to achieve better ppa (like Triton). For NPUs, the attribute can be set like what Triton does.
- Migration to tensor/linalg. However this may lead to a major refactor.

**Reply by chhzh123** [reply]
Yes, I agree with most of your points! We do want this encoding for array partitioning. It's a very useful feature that we haven't added.

For individual features you want to implement, please raise issues for "feature requests" and attach them here.

**Comment by kkkaishao** [comment]
Refactor progress update:
- Refactored `allo.pipeline` to `transform.allo.loop_pipeline`
- Refactored `allo.unroll` to `transform.allo.loop_unroll`
- Refactored `allo.unfold` to `transform.allo.loop_unfold`
   - fixed a bug that `allo.unfold` will annotate the operation with attribute "unroll" instead of "unfold".
- Refactored `allo.parallel` to `transform.allo.loop_parallel`
- Refactored `allo.bind` to `transform.allo.loop_bind`
- Refactored `allo.fuse` to `transform.allo.loop_flatten`
  - renamed the operation to better reflect its function
  - merged some logic to provide a cleaner code implemention
- Refactored `allo.compute_at` to `transform.allo.compute_at`
  - now the operation will raise an error if producer loop nest and target loop nest have no dependencies between them and producer's iteration domain is not a subset (can be equal) of consumer's (usually it's an unintended schedule). The modification aims to decrease code maintenance cost and unexpected corner cases. To handle such situation, explicitly combine loop tiling/extend transformation with it.
  - exploit store-to-load forwarding technique to remove intermediate buffers & load/stores to improve robustness.
  - with `transform.match` mechanism, we only take producer op and the target loop as inputs, no need to have three inputs anymore.
  - added a constraint that the depth of the producer loop nest must be larger than that of the target loop, otherwise may cause out-of-bounds memory access and unexpected results
  - rewrote to provide a cleaner implementation using upstream utils

**Reply by chhzh123** [reply]
Could you upstream some base facilities first? I'm afraid your PR will become too large

**Reply by kkkaishao** [reply]
The refactor is happening in separate files, so the old ones stay as they are. This won't affect any of the tool's current functions until we hook it up to the Python frontend. I think it's better to do one full PR submission right before the frontend integration.
